src/deploy/mod.rs

```diff
@@ -176,76 +234,74 @@ async fn recalculate_mode_scores(
         None => "".to_string(),
     };
 
+    let beatmap_md5s: Vec<(String,)> = sqlx::query_as(&format!(
+        "SELECT DISTINCT beatmap_md5 FROM {} WHERE completed IN (2, 3) AND play_mode = ? {}",
+        scores_table, mods_query_str,
+    ))
+    .bind(mode)
+    .fetch_all(ctx.database.get().await?.deref_mut())
+    .await?;
+
     let semaphore = Arc::new(Semaphore::new(MAX_CONCURRENT_TASKS));
 
-    let mut offset: u32 = 0;
-
-    loop {
-        let scores: Vec<RippleScore> = sqlx::query_as(
-            &format!(
-                "SELECT s.id, s.beatmap_md5, s.userid, s.score, s.max_combo, s.full_combo, s.mods, s.300_count,
-                s.100_count, s.50_count, s.katus_count, s.gekis_count, s.misses_count, s.time, s.play_mode, s.completed,
-                s.accuracy, s.pp, s.checksum, s.patcher, s.pinned, b.beatmap_id, b.beatmapset_id, b.song_name
-                FROM {} s
-                INNER JOIN
-                    beatmaps b
-                    USING(beatmap_md5)
-                WHERE
-                    completed IN (2, 3)
-                    AND play_mode = ?
-                    {}
-                ORDER BY pp DESC
-                LIMIT ? OFFSET ?",
-                scores_table,
-                mods_query_str,
-            )
-        )
-        .bind(mode)
-        .bind(BATCH_SIZE as u32)
-        .bind(offset)
-        .fetch_all(ctx.database.get().await?.deref_mut())
-        .await?;
+    let mut futures = FuturesUnordered::new();
 
-        if scores.is_empty() {
-            break;
-        }
+    log::info!(
+        beatmaps = beatmap_md5s.len(),
+        mode = mode,
+        rx = rx;
+        "Starting beatmap recalculation"
+    );
 
-        let mut futures = FuturesUnordered::new();
+    let mut beatmaps_processed = 0;
+    let total_beatmaps = beatmap_md5s.len();
 
-        for score in scores {
-            let semaphore = semaphore.clone();
-            let ctx = ctx.clone();
+    for (beatmap_md5,) in beatmap_md5s {
+        let semaphore = semaphore.clone();
+        let ctx = ctx.clone();
+        let mods_query_str = mods_query_str.clone();
 
-            let permit = semaphore.acquire_owned().await?;
+        let permit = semaphore.acquire_owned().await?;
 
-            futures.push(tokio::spawn(async move {
-                recalculate_score(score, ctx).await?;
-                drop(permit);
-                Ok::<(), anyhow::Error>(())
-            }))
-        }
+        futures.push(tokio::spawn(async move {
+            recalculate_beatmap(beatmap_md5, scores_table, mods_query_str, mode, rx, ctx).await?;
+            beatmaps_processed += 1;
 
-        while let Some(result) = futures.next().await {
-            if let Err(e) = result {
-                log::error!(
-                    error = e.to_string();
-                    "Processing score failed",
+            drop(permit);
```

Comment:
is drop() a rust builtin? i don't see where it's coming from 😳

Reply:
it is indeed a builtin, in this case its used to drop the permit thus decreasing the amount of leased permits so another task can go ahead